### PR TITLE
FMFR-260 - Allow upload of spreadsheet containing unused Building columns

### PR DESCRIPTION
**Pull request by tim-s-ccs**
> - Add the ability to import with gaps
> - Abstract some of the building validation
> - Abstract the import method loop DRY
> - Clean up method to find last complete
> - Add tests for uploading with gaps
> - Set up controller spec for spreadsheet_import
> - Add tests for the controller on create
> - Add tests for the importing method

### CODE REVIEW

File: app/services/facilities_management/spreadsheet_importer.rb [hunk 1]
```diff
@@ -484,6 +476,11 @@ def normalise_postcode(postcode)
...
 
   def service_codes
-    @procurement_array.map { |building| building[:procurement_building][:object].service_codes }.flatten.uniq
+    complete_procurement_array.map { |building| building[:procurement_building][:object].service_codes }.flatten.uniq
+  end
+
+  # An array of all the procurement buildings that are complete
+  def complete_procurement_array
+    @procurement_array.reject { |building| building[:skip] }
   end
```
Comment (iceraluk) [inline app/services/facilities_management/spreadsheet_importer.rb:485]: worth doing the following so it doesn't run the reject each time the method is called:

```
def complete_procurement_array
    @complete_procurement_array ||= @procurement_array.reject { |building| building[:skip] }
  end
```

**Review by iceraluk — APPROVED** [review]

File: app/services/facilities_management/spreadsheet_importer.rb [hunk 2]
```diff
@@ -484,6 +476,11 @@ def normalise_postcode(postcode)
...
 
   def service_codes
-    @procurement_array.map { |building| building[:procurement_building][:object].service_codes }.flatten.uniq
+    complete_procurement_array.map { |building| building[:procurement_building][:object].service_codes }.flatten.uniq
```
Comment (iceraluk) [inline app/services/facilities_management/spreadsheet_importer.rb:479]: same here 
`@service_codes ||= complete_procurement_array.map { |building| building[:procurement_building][:object].service_codes }.flatten.uniq`

**Review by iceraluk — APPROVED** [review]

File: app/services/facilities_management/spreadsheet_importer.rb (hunk 1 above)
Comment (tim-s-ccs) [reply app/services/facilities_management/spreadsheet_importer.rb:485]: I did not think to do that, I shall add these changes in

File: app/services/facilities_management/spreadsheet_importer.rb (hunk 2 above)
Comment (tim-s-ccs) [reply app/services/facilities_management/spreadsheet_importer.rb:479]: This method is only called once right at the end so it might not be necessary

**Pull request merged**